src/katsdpcontroller/product_config.py

```diff
@@ -1396,6 +1397,81 @@ def from_config(
         )
 
 
+class GpucbfTiedArrayResampledVoltageStream:
+    """Real tied-array-resampled-voltage stream (GPU V-engine)."""
+
+    stream_type: ClassVar[str] = "gpucbf.tied_array_resampled_voltage"
+    _valid_src_types: ClassVar[_ValidTypes] = {"gpucbf.tied_array_channelised_voltage"}
+
+    def __init__(
+        self,
+        name: str,
+        src_streams: Sequence[Stream],
+        *,
+        n_chans: int,
+        pols: Sequence[str],
+        station_id: str,
+    ):
+        if len(src_streams) != 2:
+            raise ValueError(f"{self.stream_type} requires two {self._valid_src_types}")
+        grandparent_streams: List[GpucbfAntennaChannelisedVoltageStream] = [
+            tacv_src_streams.src_streams[0]
+            for tacv_src_streams in src_streams
+            if len(tacv_src_streams.src_streams) == 1
+            and isinstance(tacv_src_streams.src_streams[0], GpucbfAntennaChannelisedVoltageStream)
+        ]
+        _grandparent_stream_type = "gpucbf.antenna_channelised_voltage"
+        if len(grandparent_streams) != 2:
+            # Should be two in total, one per src_stream
+            raise ValueError(
+                f"{self.stream_type} requires {self._valid_src_types} with a single src_stream"
+            )
+        if grandparent_streams[0] != grandparent_streams[1]:
+            raise ValueError(
+                f"{self._valid_src_types} src_streams need to have the same parent "
+                f"{_grandparent_stream_type} stream"
+            )
+
+        try:
+            getattr(getattr(grandparent_streams[0], "narrowband"), "vlbi")
```

Comment:
getattr with a constant string is no different from normal attribute access (other than that mypy might complain). I'd suggest you do

```suggestion
            vlbi = grandparent_streams[0].narrowband.vlbi
```

Then lower down you can just use `vlbi.pass_bandwidth` as you'll already have a GpucbfNarrowbandVLBIConfig that mypy knows is not None.

If mypy complains, an alternative check (assuming mypy is smart enough) is to say
```python
if grandparent_streams[0].narrowband is not None and grandparent_streams[0].narrowband.vlbi is not None:
    vlbi = grandparent_streams[0].narrowband.vlbi
else:
    raise ValueError(...)
```